contracts/StakingBank.sol

```diff
@@ -1,78 +1,100 @@
+  }
+
+  function contractName() external view returns (bytes32) {
+    return NAME;
+  }
+
+  function withdraw(uint256 _value)
+  external
+  onlyWhenRevealPhase
+  returns (bool) {
+    StakingBankStorage bankStorage = _storage();
+    HumanStandardToken token = HumanStandardToken(address(bankStorage.token()));
+
+    uint256 balance = bankStorage.stakingBalances(msg.sender);
 
-    Token token = Token(tokenAddress);
+    require(_value > 0 && _value <= balance, "nothing to withdraw");
 
-    require(token.transfer(msg.sender, _value));
+    require(token.transfer(msg.sender, _value), "withdraw failed");
 
-    balances[msg.sender] -= _value;
+    balance = balance.sub(_value);
+    bankStorage.setStakingBalance(msg.sender, balance);
+
+    _notifyVerifierRegistryAboutDecreasingBalance(msg.sender, _value);
 
     return true;
   }
 
-  function burn(address _id, uint256 _value) public onlyOwner nonReentrant returns (bool success) {
-    uint256 balance = balances[_id];
-    require(_value > 0 && balance >= _value);
+  // when working with `Salable.sol` token, please update this function to:
+  // `function receiveApproval(address _from)`
+  function receiveApproval(address _from, uint256 _value, address _token, bytes calldata _data)
+  external
+  returns (bool) {
+    StakingBankStorage bankStorage = _storage();
+    HumanStandardToken token = HumanStandardToken(address(bankStorage.token()));
+
+    uint256 allowance = token.allowance(_from, address(this));
+    require(allowance > 0, "nothing to approve");
 
-    Token token = Token(tokenAddress);
-    require(token.transfer(0x0, _value));
+    require(token.transferFrom(_from, address(this), allowance), "transferFrom failed");
+    bankStorage.setStakingBalance(_from, bankStorage.stakingBalances(_from).add(allowance));
 
-    balances[msg.sender] -= _value;
+    _notifyVerifierRegistryAboutIncreasingBalance(_from, allowance);
 
     return true;
   }
 
-  function lockAccount(address _id) public onlyOwner nonReentrant returns (bool success) {
-    locks[_id] = true;
-    return true;
+  function _notifyVerifierRegistryAboutIncreasingBalance(address _verifier, uint256 _add)
+  private {
+    IVerifierRegistry vr = IVerifierRegistry(contractRegistry.contractByName("VerifierRegistry"));
+    require(address(vr) != address(0x0), "VerifierRegistry address unknown");
+    require(vr.increaseShardBalance(_verifier, _add), "_notifyVerifierRegistryAboutStakeBalance failed");
   }
 
-  function unlockAccount(address _id) public onlyOwner nonReentrant returns (bool success) {
-    locks[_id] = false;
-    return true;
+  function _notifyVerifierRegistryAboutDecreasingBalance(address _verifier, uint256 _sub)
+  private {
+    IVerifierRegistry vr = IVerifierRegistry(contractRegistry.contractByName("VerifierRegistry"));
+    require(address(vr) != address(0x0), "VerifierRegistry address unknown");
+    require(vr.decreaseShardBalance(_verifier, _sub), "_notifyVerifierRegistryAboutStakeBalance failed");
```

Comment:
can we have 1 function? I'm not sure if someone still prevents passing negative values

Reply:
while developing and testing I fount this construction very helpful, because I know if its balance problem or address problem.